pkg/api/datacenter_cache.go

```diff
@@ -0,0 +1,117 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Datacenter models a single datacenter as returned by the Fastly API.
+type Datacenter struct {
+	Code  string `json:"code"` // Prometheus label is "datacenter" to make grouping at query time less tedious
+	Name  string `json:"name"`
+	Group string `json:"group"`
+}
+
+// DatacenterCache polls api.fastly.com/datacenters and maintains a local cache
+// of the returned metadata. That information is exposed as Prometheus metrics.
+type DatacenterCache struct {
+	client HTTPClient
+	token  string
+
+	mtx sync.Mutex
+	dcs []Datacenter
+}
+
+// NewDatacenterCache returns an empty cache of datacenter metadata. Use the
+// Refresh method to update the cache.
+func NewDatacenterCache(client HTTPClient, token string) *DatacenterCache {
+	return &DatacenterCache{
+		client: client,
+		token:  token,
+	}
+}
+
+// Refresh the cache with metadata retreived from the Fastly API.
+func (c *DatacenterCache) Refresh(ctx context.Context) error {
+	req, err := http.NewRequestWithContext(ctx, "GET", "https://api.fastly.com/datacenters", nil)
+	if err != nil {
+		return fmt.Errorf("error constructing API datacenters request: %w", err)
+	}
+
+	req.Header.Set("Fastly-Key", c.token)
+	req.Header.Set("Accept", "application/json")
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error executing API datacenters request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return NewError(resp)
+	}
+
+	var response []Datacenter
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return fmt.Errorf("error decoding API datacenters response: %w", err)
+	}
+
+	sort.Slice(response, func(i, j int) bool {
+		return response[i].Code < response[j].Code
+	})
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.dcs = response
+
+	return nil
+}
+
+// Datacenters returns a copy of the currently cached datacenters.
+func (c *DatacenterCache) Datacenters() []Datacenter {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	dcs := make([]Datacenter, len(c.dcs))
+	copy(dcs, c.dcs)
+	return dcs
+}
+
+// Gatherer returns a Prometheus gatherer which will yield current metadata
+// about Fastly datacenters as labels on a gauge metric.
+func (c *DatacenterCache) Gatherer(namespace, subsystem string) (prometheus.Gatherer, error) {
+	var (
+		fqName      = prometheus.BuildFQName(namespace, subsystem, "datacenter_info")
```

Comment:
Hmm, the `subsystem` makes this `fastly_rt`, rather than `fastly_api`.

Reply:
My view is that the "subsystem" of this exporter is Fastly real-time stats, and the fact that we also include some metadata from the normal API is kind of incidental. I admit in this specific case it's a bit of a gray area since the metric deals only with that normal-API-sourced information, but I guess it's good if all the metrics from this process share a common prefix.